Add get endpoint

simon0191 opened the pull request:
- Simplify router
- Change `POST /{index}` to `POST /{index}/{id}` and receive data as JSON to match elastic search API
- Add GET /{index}/{id} endpoint

simon0191 committed:
Allow create index endpoint to receive data

- Create index endpoint receives index name, id and data as JSON
- Simplify routes

simon0191 committed:
Save bleve files in .db/ folder

simon0191 committed:
Add get endpoint

0x5d commented on http/http.go:
```diff
@@ -19,34 +22,68 @@ type Server struct {
-	s := r.PathPrefix(indexPath)
-	s.HandlerFunc(createIndex(server.Index)).Methods(http.MethodPost)
+	router := mux.NewRouter()
+	router.HandleFunc("/{index}/{id}", createIndex(server.Index)).Methods(http.MethodPost)
```
This should still be prefixed by something like `/index` or `/api/index`, for namespacing.

simon0191 replied:
We should discuss if we want to be as similar as possible to elastic search. If yes, the elastic search API doesn't use an api nor index prefix in their routes, if no, I agree with you, we should prefix the routes with `/api`.

c4milo replied:
I wouldn't mind a compatibility lawyer with ElasticSearch's API

0x5d commented on http/http.go:
```diff
@@ -19,34 +22,68 @@ type Server struct {
-		splitPath := strings.Split(path, "/")
-
-		if len(splitPath) < 2 || splitPath[1] == "" {
-			http.Error(w, "No index was specified.", http.StatusBadRequest)
```
Why remove the `http.Error`?

simon0191 replied:
If there's no index specified the server will automatically respond 404 because there's no route that matches the path.

0x5d commented on http/http.go:
```diff
@@ -19,34 +22,68 @@ type Server struct {
+	}
+}
+
+func showIndex(indexService service.IndexService) http.HandlerFunc {
```
Consider renaming this to `GetIndex`, "show" is a little misleading.

simon0191 replied:
👍

c4milo replied:
It could also just be just `Index`

0x5d replied:
@c4milo That would also be ambiguous (index could also refer to the action of indexing a document).

0x5d commented on service/index.go:
```diff
@@ -1,21 +1,65 @@
+	if err == bleve.ErrorIndexPathDoesNotExist {
+		log.Printf("Creating new index " + indexName)
+		mapping := bleve.NewIndexMapping()
+		index, err := bleve.New(".db/"+indexName, mapping)
```
This should probably be configurable. Maybe not right now, tho.

simon0191 replied:
👍

c4milo replied:
`filepath.Join` is more portable.

0x5d commented on service/index.go:
```diff
@@ -1,21 +1,65 @@
+		mapping := bleve.NewIndexMapping()
+		index, err := bleve.New(".db/"+indexName, mapping)
+		if err != nil {
+			log.Fatal(err)
```
We should handle logs with [logrus](https://github.com/Sirupsen/logrus) or something more robust and configurable, to be able to disable it and have different log levels.

simon0191 replied:
👍

0x5d commented on http/http.go:
```diff
@@ -19,34 +22,68 @@ type Server struct {
-		if len(splitPath) < 2 || splitPath[1] == "" {
-			http.Error(w, "No index was specified.", http.StatusBadRequest)
+		vars := mux.Vars(r)
+		index := vars["index"]
```
This is way cleaner :smile:

simon0191 replied:
:D

0x5d commented on service/index.go:
```diff
@@ -1,21 +1,65 @@
+// Get returns a document with id `id`
+func (*DefaultIndexService) Get(indexName string, id string) (interface{}, error) {
+	//TODO: Don't open and close the index for every request
+	index, err := openOrCreateIndex(indexName)
```
I don't agree with nonexistent indices being created upon attempting to query them.

simon0191 replied:
👍

c4milo commented on http/http.go:
```diff
@@ -4,10 +4,13 @@ import (
+	"io"
+	"io/ioutil"
+	"log"
+	"reflect"
```
reflect package is only really needed when decoding or decoding protocols into Go's struct types. Other than that is usually considered non-idiomatic and in some cases a bad practice. There is some history to it already: https://codegangsta.io/blog/2014/05/19/my-thoughts-on-martini/

simon0191 committed:
Refactor

- Remove log and reflect dependencies
- Use filepath.join instead of manually create a path

simon0191 committed:
Don't create index if it doesn't exist on IndexService.Get

simon0191 committed:
Fix tests
